trackintel/preprocessing/positionfixes.py

```diff
@@ -386,3 +336,91 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
+    ret_spts = pd.DataFrame(columns=['user_id', 'started_at', 'finished_at', 'geom'])
+    
+    # pfs id should be in index
+    pfs = df.sort_values('tracked_at').reset_index().to_dict('records')
+    num_pfs = len(pfs)
+
+
+    i = 0
+    j = 0  # is zero because it gets incremented in the beginning
+    while i < num_pfs:
+        if j == num_pfs:
+            # We're at the end, this can happen if in the last "bin", 
+            # the dist_threshold is never crossed anymore.
+            break
+        else:
+            j = i + 1
+        while j < num_pfs:
+            # TODO: Can we make distance function independent of projection?
+            dist = dist_func(pfs[i][name_geocol].x, pfs[i][name_geocol].y,
+                                pfs[j][name_geocol].x, pfs[j][name_geocol].y)
+
+            if dist > dist_threshold:
+                delta_t = pfs[j]['tracked_at'] - pfs[i]['tracked_at']
+                if delta_t.total_seconds() > time_threshold:
+                    staypoint = {}
+                    staypoint['user_id'] = pfs[i]['user_id']
+                    staypoint[name_geocol] = Point(np.mean([pfs[k][name_geocol].x for k in range(i, j)]),
+                                                    np.mean([pfs[k][name_geocol].y for k in range(i, j)]))
+                    if elevation_flag:
+                        staypoint['elevation'] = np.mean([pfs[k]['elevation'] for k in range(i, j)])
+                    staypoint['started_at'] = pfs[i]['tracked_at']
+                    staypoint['finished_at'] = pfs[j - 1][
+                        'tracked_at']  # TODO: should this not be j-1? because j is not part of the staypoint. DB: Changed.
+
+                    # store matching, index should be the id of pfs
+                    staypoint['pfs_id'] = [pfs[k]['id'] for k in range(i, j)]
+
+                    # add staypoint
+                    ret_spts = ret_spts.append(staypoint, ignore_index=True)
+
+                    # TODO Discussion: Is this last point really a staypoint? As we don't know if the
+                    #      person "moves on" afterwards...
+                    if j == num_pfs - 1:
+                        staypoint = {}
+                        staypoint['user_id'] = pfs[j]['user_id']
+                        staypoint[name_geocol] = Point(pfs[j][name_geocol].x, pfs[j][name_geocol].y)
+                        if elevation_flag:
+                            staypoint['elevation'] = pfs[j]['elevation']
+                        staypoint['started_at'] = pfs[j]['tracked_at']
+                        staypoint['finished_at'] = pfs[j]['tracked_at']
+                        # store matching, index should be the id of pfs
+                        staypoint['pfs_id'] = [pfs[j]['id']]
+
+                        ret_spts = ret_spts.append(staypoint, ignore_index=True)
+                i = j
+                break
+            j = j + 1
+    
+    return ret_spts
+
+def _generate_staypoints_dbscan_user(df, 
```

Comment:
I think this function should get the metric parameter from the main function.

Reply:
Yes, I agree. In the previous version the DBSCAN just used L2 distance and ignores the distance function. This will be fixed in the fix DBSCAN pull request